### PR TITLE
Timeout Overrides in Messaging Templates

### PR DESCRIPTION
**Pull request by garyrussell**
Issue: SPR-15591

Provide a mechanism to override the configured send and receive timeouts in the
GenericMessagingTemplate.

- overload `doReceive()` to take a receive timeout argument
- for `sendAndReceive()` methods examine message headers for these timeout values
  - remove headers to avoid propagation
  - avoid the unconditional rebuild of the reply message if the original headers weren't present
- also remove headers from simple `send()` operations
- change javadocs for the setters to indicate they are now defaults
- add properties to allow the user to override the header names used
- change `TemporaryReplyChannel` to use the `send` argument and change it to static
  - add package-protected ctor to avoid the compiler creating a synthetic ctor for access

### CODE REVIEW

**Comment by artembilan** [comment]
> JUnit tests should include assert() or fail()

Looks like `ArgumentCaptor` and `assert...` afterwards should be used instead of asserts in the `Answer`

**Comment by artembilan** [comment]
> JUnit tests should include assert() or fail()

Looks like `ArgumentCaptor` and `assert...` afterwards should be used instead of asserts in the `Answer`

File: spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java [hunk 1]
```diff
@@ -82,6 +96,47 @@ public long getReceiveTimeout() {
...
+		Assert.notNull(sendTimeoutHeader, "'sendTimeoutHeader' cannot be null");
+		this.sendTimeoutHeader = sendTimeoutHeader;
+	}
+
+	/**
+	 * @return the configured sendTimeoutHeader.
+	 * @since 5.0
+	 */
+	public String getSendTimeoutHeader() {
+		return sendTimeoutHeader;
```
Comment (artembilan) [inline spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java]: ` this.` ?

File: spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java [hunk 2]
```diff
@@ -82,6 +96,47 @@ public long getReceiveTimeout() {
...
+		Assert.notNull(receiveTimeoutHeader, "'receiveTimeoutHeader' cannot be null");
+		this.receiveTimeoutHeader = receiveTimeoutHeader;
+	}
+
+	/**
+	 * @return the configured receiveTimeoutHeader
+	 * @since 5.0
+	 */
+	public String getReceiveTimeoutHeader() {
+		return receiveTimeoutHeader;
```
Comment (artembilan) [inline spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java]: ` this.`, too

**Review by artembilan — CHANGES_REQUESTED** [review]
Just a couple concerns from me.

**Comment by rstoyanchev** [comment]
This is merged into master as a single commit. Thanks Gary.

File: spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java [hunk 3]
```diff
@@ -142,20 +212,25 @@ protected final void doSend(MessageChannel channel, Message<?> message) {
...
 		}
 		catch (RuntimeException ex) {
 			tempReplyChannel.setSendFailed(true);
 			throw ex;
 		}
 
-		Message<?> replyMessage = this.doReceive(tempReplyChannel);
-		if (replyMessage != null) {
+		Message<?> replyMessage = this.doReceive(tempReplyChannel, receiveTimeout);
+		if (replyMessage != null && (originalReplyChannelHeader!= null || originalErrorChannelHeader != null)) {
```
Comment (garyrussell) [inline spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java]: @rstoyanchev It turns out this change was incorrect; we need to null-out the channel headers if they were originally null.

Do you want me to submit another PR, or would you mind just reverting this line?

Thanks (and my apologies).

Comment (rstoyanchev) [inline spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java]: If you could push another commit on the same branch, so I can see the exact changes you have in mind.

Comment (garyrussell) [inline spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java]: Sorry - I just issued a new PR on that branch (having first reset it to master).

I issued another PR: https://github.com/spring-projects/spring-framework/pull/1440

Thanks

**Pull request closed without merging**